packages/engine/Source/Widget/CesiumWidget.js

```diff
@@ -829,4 +1136,512 @@ CesiumWidget.prototype.render = function () {
+    };
+  });
+  that._zoomPromise = zoomPromise;
+  that._zoomIsFlight = isFlight;
+  that._zoomOptions = options;
+
+  Promise.resolve(zoomTarget).then(function (zoomTarget) {
+    //Only perform the zoom if it wasn't cancelled before the promise resolved.
+    if (that._zoomPromise !== zoomPromise) {
+      return;
+    }
+
+    //If the zoom target is a rectangular imagery in an ImageLayer
+    if (zoomTarget instanceof ImageryLayer) {
+      let rectanglePromise;
+
+      if (defined(zoomTarget.imageryProvider)) {
+        rectanglePromise = Promise.resolve(zoomTarget.getImageryRectangle());
+      } else {
+        rectanglePromise = new Promise((resolve) => {
+          const removeListener = zoomTarget.readyEvent.addEventListener(() => {
+            removeListener();
+            resolve(zoomTarget.getImageryRectangle());
+          });
+        });
+      }
+      rectanglePromise
+        .then(function (rectangle) {
+          return computeFlyToLocationForRectangle(rectangle, that.scene);
+        })
+        .then(function (position) {
+          //Only perform the zoom if it wasn't cancelled before the promise was resolved
+          if (that._zoomPromise === zoomPromise) {
+            that._zoomTarget = position;
+          }
+        });
+      return;
+    }
+
+    if (
+      zoomTarget instanceof Cesium3DTileset ||
+      zoomTarget instanceof TimeDynamicPointCloud ||
+      zoomTarget instanceof VoxelPrimitive
+    ) {
+      that._zoomTarget = zoomTarget;
+      return;
+    }
+
+    //If the zoom target is a data source, and it's in the middle of loading, wait for it to finish loading.
+    if (zoomTarget.isLoading && defined(zoomTarget.loadingEvent)) {
+      const removeEvent = zoomTarget.loadingEvent.addEventListener(function () {
+        removeEvent();
+
+        //Only perform the zoom if it wasn't cancelled before the data source finished.
+        if (that._zoomPromise === zoomPromise) {
+          that._zoomTarget = zoomTarget.entities.values.slice(0);
+        }
+      });
+      return;
+    }
+
+    //Zoom target is already an array, just copy it and return.
+    if (Array.isArray(zoomTarget)) {
+      that._zoomTarget = zoomTarget.slice(0);
+      return;
+    }
+
+    //If zoomTarget is an EntityCollection, this will retrieve the array
+    zoomTarget = defaultValue(zoomTarget.values, zoomTarget);
+
+    //If zoomTarget is a DataSource, this will retrieve the array.
+    if (defined(zoomTarget.entities)) {
+      zoomTarget = zoomTarget.entities.values;
+    }
+
+    //Zoom target is already an array, just copy it and return.
+    if (Array.isArray(zoomTarget)) {
+      that._zoomTarget = zoomTarget.slice(0);
+    } else {
+      //Single entity
+      that._zoomTarget = [zoomTarget];
+    }
+  });
+
+  that.scene.requestRender();
+  return zoomPromise;
+}
+
+function clearZoom(viewer) {
+  viewer._zoomPromise = undefined;
+  viewer._zoomTarget = undefined;
+  viewer._zoomOptions = undefined;
+}
```

Comment:
For this and a handful of functions below that were moved over the argument names should be swapped out so they don't refer to a `viewer` anymore
```suggestion
function clearZoom(widget) {
  widget._zoomPromise = undefined;
  widget._zoomTarget = undefined;
  widget._zoomOptions = undefined;
}
```

Reply:
Fixed

Reply:
@jjspace Thanks for the review. I agree, this is not very easy to track the changes when things move around between files, whatever the tool...
Following your suggestion, I've removed the viewer's tests that are now done in the widget's spec. Unsurprisingly, the coverage decreases for the former, while it increases for the latter...
Tell me if you see anything else!